### PR TITLE
Fixing error for using TSS with tboot 1.8.3(TPM 2.0)

### PR DESCRIPTION
**Pull request by kkamagui**
This commit solves issue #83. tboot 1.8.3(TPM 2.0) creates a primary hierarchy handle and an object handle during boot process. However the resource manager of TSS seems to presume these handles as abnormal objects and returns an error code to the client.

To solve this problem, the resource manager is changed to show warning instead of returning an error to the client.

### CODE REVIEW

**Comment by kkamagui** [comment]
Hello, wcarthur1.

Thank you for the reply.

I am trying to enable tboot 1.8.3 with NUC5i5MYHE utilizing your TPM2.0-TSS, and it should run like,

1. Boot linux using tboot 1.8.3
2. Start the resource manager of TSS
3. Generate random numbers using tpm2_getrandom of sapi-tools in tpm2.0-tools

When I try to get a random number, the tpm2_getrandom of sapi-tools reports an error. It seems the error is due to responseRval(line 2066 in the RM.c) has TSS2_RESMGR_UNLOADED_OBJECTS, because testForLoadedSessionsOrObjectsRval gets TSS2_RESMGR_UNLOADED_OBJECTS(the failure) at line 2061 in the RM.c

However, this behavior doesn’t look right. I guess TSS should manage only the objects and the sessions that are created after TSS itself is started, and may not recognize the objects and the sessions that are created before it is started.

In other words, when linux is loaded, tboot creates an object handle and the RM doesn't know about this object. Consequently, the RM returns an error code to the client. If the RM is terminated abnormally due to an exception, the RM may not flush the objects and the sessions, and the restarted RM may have a same problem.

```bash
# Log messages from tboot 1.8.3
TBOOT: TPM:CreatePrimary creating hierarchy handle = 40000007
TBOOT: TPM:CreatePrimary created object handle = 80000000
```

In my opinion, the RM needs to identify the objects and the sessions that are created prior to the RM starts, and the RM needs to flush them.

Please let me know if I understand the codes wrong, and please let me show my new solution of this issue.

Thank you.

**Comment by wcarthur1** [comment]
Below…

BTW, who do you work for?  Do you work for Intel?  If so, what group are you in?

Will Arthur
Intel Corporation
Server Security Firmware
803-216-2101

**Comment by kkamagui** [comment]
Hello, wcarthur1.

Thank you for the reply.

I work for a security research institute of government. My team has designed a security platform for PC and I think that TPM 2.0 will be a core part of the security platform. So I am trying to enable tboot 1.8.3 with NUC5i5MYHE utilizing TPM2.0-TSS.

My test machine is booted using tboot 1.8.3, and tboot creates an object handle during boot process. I could find this object handle(0x80000000) in tboot logs(using txt-state command). 
```bash
# Log messages from tboot 1.8.3
TBOOT: TPM: TPM Family 0x1
TBOOT: TPM: supported bank count = 2
TBOOT: TPM: bank alg = 00000004
TBOOT: TPM: bank alg = 0000000b
TBOOT: TPM: supported alg cout = 00000002
TBOOT: 		 00000004
TBOOT: 		 0000000B
TBOOT: TPM:CreatePrimary creating hierarchy handle = 40000007
TBOOT: TPM:CreatePrimary created object handle = 80000000              <== object handle for tboot
TBOOT: TPM attribute:
TBOOT: 	 extend policy 2
TBOOT: 	 current alg id 0x4
TBOOT: 	 timeout values: A: 750, B: 2000, C: 75000, D: 750
```

When I generate random numbers using tpm2_getrandom of sapi-tools in tpm2.0-tools, the RM dumps following error logs related to this object handle.
```bash
# The RM’s log
TPM2.0-TSS/resourcemgr$> ./resourcemgr 
Initializing local TPM Interface
Initializing Resource Manager
maxActiveSessions = 32
gapMaxValue = 255
socket created:  0x4
bind to IP address:port:  127.0.0.1:2324
Other CMD server listening to socket:  0x4
socket created:  0x5
bind to IP address:port:  127.0.0.1:2323
TPM CMD server listening to socket:  0x5
Starting SockServer (TPM CMD), socket: 0x5.
Starting SockServer (Other CMD), socket: 0x4.

Accept socket:  0x7
Resource Manager Other CMD Server accepted client
Accept socket:  0x6
Resource Manager TPM CMD Server accepted client
||  Loaded transient object handles: 
||  0x80000000, TpmCmdServer died (TPM CMD), rval: 0x00000000, socket: 0x6.
OtherCmdServer died (Other CMD), socket: 0x7.
```

```bash
# The tpm2_getrandom ’s log
tpm2.0-tools/sapi-tools$> ./tpm2_getrandom -s 10 -o a.txt
......TPM2_GetRandom Error. TPM Error:0xc6000......
```

I agree with your ideas and I think that y

our 2nd choice is better than others. As I mentioned in previous comment, if the RM is terminated abnormally due to an unexpected exception, the RM may not flush the objects that the RM has. In this case, the restarted RM also returns an error code because the new RM doesn’t know about previous handles. In my opinion, we should assume that the state before starting the RM is not always clean, and the RM needs to flush existing handles for robustness.

So I changed codes that the RM flushes all object handles when the RM is started, and it worked fine for all cases(with and without tboot). 
If you don’t mind, please review attached patch codes.

Thank you. 

[This-commit-solves-issue-83.patch.txt](https://github.com/01org/TPM2.0-TSS/files/57626/This-commit-solves-issue-83.patch.txt)

**Comment by kkamagui** [comment]
Hello, wcarthur1.

Thank you for the replay.

For attaching the RM's logs, I wrote a comment to the github.

Here is the link:
https://github.com/01org/TPM2.0-TSS/pull/85#issuecomment-163484218

Thank you.

Seunghun Han.

**Comment by wcarthur1** [comment]
Below…

I’m going to add one of our TBOOT contacts here…

Ning, do you have a recommendation here?

Will Arthur
Intel Corporation
Server Security Firmware
803-216-2101

**Comment by wcarthur1** [comment]
[Arthur, Will C] How is TBOOT using this object?  Does it need to use it later?

[Ning] this is primary object created by tboot to create keys to seal tboot mem in a S3 resume, and unseal tboot mem after return from S3 resume.

Ning Sun

Intel Corporation



Tboot project: - http://sourceforge.net/projects/tboot/

Intel® TXT MLE DG: http://download.intel.com/technology/security/downloads/315168.pdf



From: Arthur, Will C
Sent: Thursday, December 10, 2015 5:29 AM
To: 01org/TPM2.0-TSS; 01org/TPM2.0-TSS; Sun, Ning
Cc: Wiseman, Monty; Sun, Ning
Subject: RE: [TPM2.0-TSS] Fixing error for using TSS with tboot 1.8.3(TPM 2.0) (#85)

Below…

I’m going to add one of our TBOOT contacts here…

Ning, do you have a recommendation here?

Will Arthur
Intel Corporation
Server Security Firmware
803-216-2101

**Comment by wcarthur1** [comment]
Ning,

For this to be guaranteed to work correctly, TBOOT needs to start the resource manager before creating this key.

Otherwise, it will not be able to seal TBOOT before S3 and after S3 resume.

Will Arthur
Intel Corporation
Server Security Firmware
803-216-2101

From: Sun, Ning
Sent: Thursday, December 10, 2015 1:40 PM
To: Arthur, Will C <will.c.arthur@intel.com>; 01org/TPM2.0-TSS <reply@reply.github.com>; 01org/TPM2.0-TSS <TPM2.0-TSS@noreply.github.com>
Cc: Wiseman, Monty <monty.wiseman@intel.com>
Subject: RE: [TPM2.0-TSS] Fixing error for using TSS with tboot 1.8.3(TPM 2.0) (#85)


[Arthur, Will C] How is TBOOT using this object?  Does it need to use it later?

[Ning] this is primary object created by tboot to create keys to seal tboot mem in a S3 resume, and unseal tboot mem after return from S3 resume.

Ning Sun

Intel Corporation



Tboot project: - http://sourceforge.net/projects/tboot/

Intel® TXT MLE DG: http://download.intel.com/technology/security/downloads/315168.pdf



From: Arthur, Will C
Sent: Thursday, December 10, 2015 5:29 AM
To: 01org/TPM2.0-TSS; 01org/TPM2.0-TSS; Sun, Ning
Cc: Wiseman, Monty; Sun, Ning
Subject: RE: [TPM2.0-TSS] Fixing error for using TSS with tboot 1.8.3(TPM 2.0) (#85)

Below…

I’m going to add one of our TBOOT contacts here…

Ning, do you have a recommendation here?

Will Arthur
Intel Corporation
Server Security Firmware
803-216-2101

**Comment by kkamagui** [comment]
wcarthur1 and Ning.


The object handle that Ning mentioned(handle2048 variable at tpm_20.c in
tboot source) is used for entering into S3 state and resuming from S3
state. But the object handle seems to be created from TPM_RH_NULL, which
means that the object handle could be created anytime when tboot needs.


I would like to suggest that tboot creates/flushes the object handle when
enters into/resumes from the S3 state. The patch that I sent
previously(FlushAllLoadedHandles() function) should work O.K, and it is
needed for robustness of the RM. It would not interfere tboot working, as
long as it gets/returns the handle whenever needs.


Ning, please let me know if I understand wrong. Otherwise, I could add some
modification to tpm_20.c file, so that tboot could create the object handle
as necessary.


Thank you.

**Comment by wcarthur1** [comment]
Ning and I met yesterday and agreed on some TBOOT changes.  Until those are incorporated into TBOOT, you will not be able to successfully S3 and resume.

As for your changes I like the change to flush any objects in the RM initialization.  I would like to suggest the following change:  please change the warning at the end of the RM receive call to an error like it was before.  Because this can cause problems that are hard to debug and confusing, I would this to be a very visible error.

Will Arthur
Intel Corporation
Server Security Firmware
803-216-2101

**Comment by kkamagui** [comment]
Hello, wcarthur1.

That’s good news to me. Thank you very much.

According to your suggestion, I will change the warning to the error like it was before and close this pull-request.

If you don’t mind, I would like to make new pull-request that previous patch code([This-commit-solves-issue-83.patch.txt](https://github.com/01org/TPM2.0-TSS/files/57626/This-commit-solves-issue-83.patch.txt)) is applied. The patch has the codes that flush any objects in the RM initialization, and it works fine with/without tboot.

Please allow me to make new pull-request and review it.

Thank you.

**Comment by wcarthur1** [comment]
Yes, that sounds good.

And thanks for helping with this issue!

Will Arthur
Intel Corporation
Server Security Firmware
803-216-2101

**Comment by kkamagui** [comment]
Hello, wcarthur1.


That’s good news to me. Thank you very much.


According to your suggestion, I will change the warning to the error like
it was before and close this pull-request.


If you don’t mind, I would like to make new pull-request that previous
patch code(
https://github.com/01org/TPM2.0-TSS/files/57626/This-commit-solves-issue-83.patch.txt)
is applied. The patch has the codes that flush any objects in the RM
initialization, and it works fine with/without tboot.


Please allow me to make new pull-request and review it.


Thank you.

**Comment by wcarthur1** [comment]
I would like to review the pull request after you create it.  Always helps to have another set of eyes on the code.

Will Arthur
Intel Corporation
Server Security Firmware
803-216-2101

**Comment by kkamagui** [comment]
Hello, wcarthur1.

Following your suggestion, I made a new pull-request at
https://github.com/01org/TPM2.0-TSS/pull/86 .
Please review it and if you have any comment, please let me know.

Thank you.

========
Seunghun Han
National Security Research Institute
+82-42-870-4781
